### PR TITLE
Document seemingly unneeded self.clone in write

### PR DESCRIPTION
**Pull request by krtab**
I don't think the clone is needed.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -116,20 +116,19 @@ impl Hasher for FxHasher {
         const _: () = assert!(size_of::<usize>() <= size_of::<u64>());
         // Ensure no bytes are discarded by casting to usize
         const _: () = assert!(size_of::<u32>() <= size_of::<usize>());
-        let mut state = self.clone();
```
Comment (compiler-errors) [inline src/lib.rs:119]: @calebsander: Why did you add this clone in #12?

Comment (krtab) [reply src/lib.rs:119]: In case they don't answer, my two guesses are:
1. It is a leftover of the refactoring that is indeed superfluous
2. It is supposed to prevent committing "partially hashed" slices of bytes to the hasher in case of a panic in `add_to_hash`. But a) I don't see currently a way for this to panic and b) I don't think it would make much sense to try to recover from a panic anyway, especially as this invariant is not documented anywhere.

Thanks for reviewing

Comment (calebsander) [reply src/lib.rs:119]: Please check the generated assembly. If I remember correctly, cloning (which is a single 64-bit load, and later a 64-bit store to save the state) allowed the hash accumulation to happen in registers. Accumulating directly on `self` meant the bitwise operations and multiplication were applied to a memory location. I figured avoiding memory accesses in hot loops (even if they are very cache-friendly) was a good idea. But it's also likely that the `Hasher` methods will be inlined anyways, with the caller storing the hash state in a local variable, so the state may well be in a register rather than memory in the first place.
It's entirely possible that the compiler no longer behaves this way, or that it doesn't really matter for performance. Adding a benchmark to give some concrete data would be much better than my speculating.

**Comment by krtab** [comment]
@calebsander Thanks for the explanation. It seems that the codegen is indeed suboptimal without the clone. I've added a comment explaining why the clone is here and opened https://github.com/rust-lang/rust/issues/123129 about the suboptimal codegen.

**Review by calebsander — APPROVED** [review]

**Review by WaffleLapkin — APPROVED** [review]
